clients/src/test/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImplTest.java

```diff
@@ -2302,6 +2303,61 @@ public void testRebalanceMetricsOnFailedRebalance() {
         assertEquals(-1d, getMetricValue(metrics, rebalanceMetricsManager.lastRebalanceSecondsAgo));
     }
 
+    @Test
+    public void testLeaveGroupWhenStateIsFatal() {
+        MembershipManagerImpl membershipManager = createMemberInStableState(null);
+        when(subscriptionState.hasAutoAssignedPartitions()).thenReturn(true);
+        membershipManager.transitionToFatal();
+        assertEquals(MemberState.FATAL, membershipManager.state());
+
+        subscriptionState.assignFromUser(Collections.singleton(new TopicPartition("topic", 0)));
+        assertEquals(1, subscriptionState.numAssignedPartitions());
```

Comment:
I'm not sure we would ever be able to achieve this path we're mocking here with a real consumer, so wondering if this is even a valid scenario to test? (this test, and the ones below for fenced and stale too).  

A member in fatal/fenced/stale state means it was part of a group (automatic assignment after subscribe), and then we're mocking a manual assignment and takes effect and is cleared on leaveGroup. I expect that would never be possible because the consumer does not allow to mix the subscription types, so the call to assign would fail before actually updating the assignment and numAssignedPartitions (a consumer would have to unsubscribe or assign with empty to then be able to manually assign the topic0 partition) 

Just for the record, I totally see the value in the `testLeaveGroupWhenStateIsUnsubscribe`, which is covering the gap we had and it's a valid combination since it's not mixing subscription types, but I struggle to see this call to assign happening when a member is fatal/fenced/stale?

Reply:
Yeah, I agree we should not have test cases which are not real cases. I change `spy` back to `mock` and only test `subscriptionState#unsubscribe` is called when `isNotInGroup` is true.

Reply:
Cool. Related although not introduced by this PR, this makes me notice that we don't have this full story covered in the tests (not testing the call to `subscriptionState.unsubscribed` when `isNotInGroup` is false).  

Would you mind adding the same verification you added for `notInGroup` true, to the func that covers this for `notInGroup` false? Should be the same `verify(subscriptionState).unsubscribe();` but added to `assertStaleMemberLeavesGroupAndClearsAssignment` I would say.

Reply:
Thanks for the suggestion. I add `verify(subscriptionState).unsubscribe();` to test cases which calling `leaveGroup`. The only different one is `testLeaveGroupWhenMemberAlreadyLeaving`. I use `verify(subscriptionState, never()).unsubscribe();` for second leave attempt. Could you take a look again? Thank you.